File: python/csrc/algorithm.cpp
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#include <nanobind/nanobind.h>
#include <nanobind/stl/function.h>
#include <nanobind/stl/pair.h>
#include <nanobind/stl/shared_ptr.h>
#include <nanobind/stl/string.h>
#include <nanobind/stl/unordered_map.h>
#include <nanobind/stl/vector.h>

#include <cstring>
#include <mscclpp/algorithm.hpp>

namespace nb = nanobind;
using namespace mscclpp;

void register_algorithm(nb::module_& m) {
  nb::enum_<CollectiveBufferMode>(m, "CollectiveBufferMode")
      .value("ANY", CollectiveBufferMode::Any)
      .value("IN_PLACE", CollectiveBufferMode::InPlace)
      .value("OUT_OF_PLACE", CollectiveBufferMode::OutOfPlace);

  nb::enum_<AlgorithmType>(m, "AlgorithmType").value("NATIVE", AlgorithmType::Native).value("DSL", AlgorithmType::DSL);

  nb::enum_<CommResult>(m, "CommResult")
      .value("COMM_SUCCESS", CommResult::CommSuccess)
      .value("COMM_UNHANDLED_CUDA_ERROR", CommResult::CommUnhandledCudaError)
      .value("COMM_SYSTEM_ERROR", CommResult::CommSystemError)
      .value("COMM_INTERNAL_ERROR", CommResult::CommInternalError)
      .value("COMM_INVALID_ARGUMENT", CommResult::CommInvalidArgument)
      .value("COMM_INVALID_USAGE", CommResult::CommInvalidUsage)
      .value("COMM_REMOTE_ERROR", CommResult::CommRemoteError)
      .value("COMM_IN_PROGRESS", CommResult::CommInProgress)
      .value("COMM_NUM_RESULTS", CommResult::CommNumResults);

  nb::enum_<ReduceOp>(m, "ReduceOp")
      .value("SUM", ReduceOp::SUM)
      .value("MIN", ReduceOp::MIN)
      .value("NOP", ReduceOp::NOP);

  auto algorithmClass =
      nb::class_<Algorithm>(m, "Algorithm")
          .def_static(
              "from_native_capsule",
              [](nb::capsule cap) {
                const char* name = cap.name();
                if (name == nullptr || std::strcmp(name, ALGORITHM_NATIVE_CAPSULE_NAME) != 0) {
                  throw nb::type_error("Invalid capsule: expected 'mscclpp::AlgorithmPtr'");
                }
                void* data = cap.data();
                if (data == nullptr) {
                  throw nb::value_error("Failed to get pointer from capsule");
                }
                return *static_cast<std::shared_ptr<Algorithm>*>(data);
              },
              nb::arg("capsule"))
          .def_prop_ro("name", &Algorithm::name)
          .def_prop_ro("collective", &Algorithm::collective)
          .def_prop_ro("message_range", &Algorithm::messageRange)
          .def_prop_ro("tags", &Algorithm::tags)
          .def_prop_ro("buffer_mode", &Algorithm::bufferMode)
          .def_prop_ro("constraint", &Algorithm::constraint)
          .def_prop_ro("type", &Algorithm::type)
          .def(
              "execute",
              [](Algorithm& self, std::shared_ptr<Communicator> comm, uintptr_t input, uintptr_t output,
                 size_t inputSize, size_t outputSize, DataType dtype, ReduceOp op, uintptr_t stream,
                 std::shared_ptr<Executor> executor, int nBlocks, int nThreadsPerBlock,
                 std::unordered_map<std::string, uintptr_t> extras) {
                return self.execute(comm, reinterpret_cast<const void*>(input), reinterpret_cast<void*>(output),
                                    inputSize, outputSize, dtype, op, reinterpret_cast<cudaStream_t>(stream), executor,
                                    nBlocks, nThreadsPerBlock, extras);
              },
              nb::arg("comm"), nb::arg("input"), nb::arg("output"), nb::arg("input_size"), nb::arg("output_size"),
              nb::arg("dtype"), nb::arg("op") = ReduceOp::NOP, nb::arg("stream") = 0, nb::arg("executor") = nullptr,
              nb::arg("n_blocks") = 0, nb::arg("n_threads_per_block") = 0,
              nb::arg("extras") = std::unordered_map<std::string, uintptr_t>())
            .def("reset", &Algorithm::reset);

  nb::class_<Algorithm::Constraint>(algorithmClass, "Constraint")
      .def(nb::init<>())
      .def(nb::init<int, int>(), nb::arg("world_size"), nb::arg("n_ranks_per_node"))
      .def_rw("world_size", &Algorithm::Constraint::worldSize)
      .def_rw("n_ranks_per_node", &Algorithm::Constraint::nRanksPerNode);

  nb::class_<AlgorithmBuilder>(m, "AlgorithmBuilder").def("build", &AlgorithmBuilder::build);

  nb::class_<DslAlgorithm, Algorithm>(m, "DslAlgorithm")
      .def(nb::init<std::string, ExecutionPlan, std::unordered_map<std::string, uint64_t>, Algorithm::Constraint>(),
           nb::arg("id"), nb::arg("plan"), nb::arg("tags") = std::unordered_map<std::string, uint64_t>(),
           nb::arg("constraint") = Algorithm::Constraint())
      .def("build", &DslAlgorithm::build);

  nb::class_<AlgorithmCollection>(m, "AlgorithmCollection")
      .def("register_algorithm", &AlgorithmCollection::registerAlgorithm, nb::arg("collective"), nb::arg("algo_name"),
           nb::arg("algorithm"))
      .def("get_algorithms_by_collective", &AlgorithmCollection::getAlgorithmsByCollective, nb::arg("collective"))
      .def("to_list", &AlgorithmCollection::getAllAlgorithms);

  nb::class_<CollectiveRequest>(m, "CollectiveRequest")
      .def_ro("world_size", &CollectiveRequest::worldSize)
      .def_ro("n_ranks_per_node", &CollectiveRequest::nRanksPerNode)
      .def_ro("rank", &CollectiveRequest::rank)
      .def_prop_ro("input_buffer",
                   [](const CollectiveRequest& self) { return reinterpret_cast<uintptr_t>(self.inputBuffer); })
      .def_prop_ro("output_buffer",
                   [](const CollectiveRequest& self) { return reinterpret_cast<uintptr_t>(self.outputBuffer); })
      .def_ro("message_size", &CollectiveRequest::messageSize)
      .def_prop_ro("collective", [](const CollectiveRequest& self) { return self.collective; })
      .def_ro("dtype", &CollectiveRequest::dtype)
      .def_prop_ro("hints", [](const CollectiveRequest& self) { return self.hints; })
      .def("buffer_mode", &CollectiveRequest::bufferMode);
}

File: python/mscclpp/_core/algorithm.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from __future__ import annotations
from typing import Optional, Tuple, Dict
from functools import cached_property


from mscclpp._mscclpp import (
    Algorithm as _Algorithm,
    DslAlgorithm as _DslAlgorithm,
    AlgorithmType as _AlgorithmType,
    Communicator,
    CollectiveBufferMode,
    DataType,
    Executor,
    ExecutionPlan,
    ReduceOp,
)

__all__ = ["Algorithm", "AlgorithmBuilder", "AlgorithmCollection"]


class Algorithm:
    """A wrapper for collective communication algorithms.

    This class provides a Python interface for collective communication algorithms
    such as allreduce, allgather, and reduce-scatter. Algorithms can be either
    DSL-based (defined using MSCCL++ execution plans) or native (implemented in C++/CUDA).

    Attributes:
        name: Human-readable name of the algorithm.
        collective: The collective operation this algorithm implements (e.g., "allreduce").
        message_size_range: Tuple of (min_size, max_size) in bytes for valid message sizes.
        tags: Dictionary of tag names to tag values for algorithm selection hints.
        buffer_mode: The buffer mode supported by this algorithm (IN_PLACE, OUT_OF_PLACE, or ANY).
    """

    class Constraint:
        """Constraints that define valid execution environments for the algorithm.

        Args:
            world_size: Required world size (number of ranks). 0 means any size.
            n_ranks_per_node: Required number of ranks per node. 0 means any.
        """

        def __init__(self, world_size: int = 0, n_ranks_per_node: int = 0):
            self._constraint = _Algorithm.Constraint(world_size, n_ranks_per_node)

        @property
        def world_size(self) -> int:
            return self._constraint.worldSize

        @property
        def n_ranks_per_node(self) -> int:
            return self._constraint.nRanksPerNode

    def __init__(
        self,
        id: Optional[str] = None,
        execution_plan: Optional[ExecutionPlan] = None,
        native_handle: Optional[_Algorithm] = None,
        tags: Optional[Dict[str, int]] = None,
        constraint: Optional[Constraint] = None,
    ):
        if execution_plan is not None:
            self._algorithm = _DslAlgorithm(
                id,
                execution_plan,
                tags=tags if tags is not None else {},
                constraint=constraint._constraint if constraint is not None else _Algorithm.Constraint(),
            )
        elif native_handle is not None:
            self._algorithm = native_handle

    @classmethod
    def create_from_native_handle(cls, handle: _Algorithm):
        """Create an Algorithm instance from a native C++ algorithm handle.

        Args:
            handle: The native C++ algorithm handle.

        Returns:
            A new Algorithm instance wrapping the native handle.
        """
        return cls(
            native_handle=handle,
        )

    @classmethod
    def create_from_native_capsule(cls, obj):
        """Create an Algorithm instance from a PyCapsule object.

        Args:
            obj: A PyCapsule containing a native algorithm pointer.

        Returns:
            A new Algorithm instance wrapping the algorithm from the capsule.
        """
        handle = _Algorithm.from_native_capsule(obj)
        return cls(native_handle=handle)

    @cached_property
    def name(self) -> str:
        """The human-readable name of the algorithm."""
        return self._algorithm.name

    @cached_property
    def collective(self) -> str:
        """The collective operation this algorithm implements (e.g., "allreduce", "allgather")."""
        return self._algorithm.collective

    @cached_property
    def message_size_range(self) -> Tuple[int, int]:
        """The valid message size range (min_size, max_size) in bytes."""
        return (self._algorithm.message_range[0], self._algorithm.message_range[1])

    @cached_property
    def tags(self) -> Dict[str, int]:
        """Dictionary of tag names to tag values for algorithm selection hints."""
        return self._algorithm.tags

    @cached_property
    def buffer_mode(self) -> CollectiveBufferMode:
        """The buffer mode supported by this algorithm (IN_PLACE, OUT_OF_PLACE, or ANY)."""
        return self._algorithm.buffer_mode

    def is_dsl_algorithm(self) -> bool:
        """Check if this is a DSL-based algorithm.

        Returns:
            True if this algorithm is defined using DSL/execution plan, False otherwise.
        """
        if self._algorithm.type == _AlgorithmType.DSL:
            return True
        return False

    def is_native_algorithm(self) -> bool:
        """Check if this is a native C++/CUDA algorithm.

        Returns:
            True if this algorithm is implemented natively, False otherwise.
        """
        if self._algorithm.type == _AlgorithmType.NATIVE:
            return True
        return False

    def execute(
        self,
        comm: Communicator,
        input_buffer: int,
        output_buffer: int,
        input_size: int,
        output_size: int,
        dtype: DataType,
        op: ReduceOp = ReduceOp.NOP,
        stream: int = 0,
        executor: Optional[Executor] = None,
        nblocks=0,
        nthreads_per_block=0,
        extras: Optional[Dict[str, int]] = None,
    ) -> int:
        """Execute the collective algorithm.

        Args:
            comm: The communicator to use.
            input_buffer: Device pointer to the input buffer.
            output_buffer: Device pointer to the output buffer.
            input_size: Size of the input buffer in bytes.
            output_size: Size of the output buffer in bytes.
            dtype: Data type of the elements.
            op: Reduction operation for reduce-type collectives (default: NOP).
            stream: CUDA stream to execute on (default: 0).
            executor: The executor for DSL algorithms (required for DSL, optional for native).
            nblocks: Number of CUDA blocks (0 for auto-selection).
            nthreads_per_block: Number of threads per block (0 for auto-selection).
            extras: Additional algorithm-specific parameters.

        Returns:
            The result code (0 for success).
        """
        return self._algorithm.execute(
            comm,
            int(input_buffer),
            int(output_buffer),
            input_size,
            output_size,
            dtype,
            op,
            int(stream),
            executor,
            nblocks,
            nthreads_per_block,
            extras if extras is not None else {},
        )

class AlgorithmBuilder:
    def __init__(self, algorithm_builder: _AlgorithmBuilder):
        self._algorithm_builder = algorithm_builder

    def build(self) -> Algorithm:
        return Algorithm.create_from_native_handle(self._algorithm_builder.build())


class AlgorithmCollection:
    def __init__(self, native_collection: _AlgorithmCollection):
        self._native_collection = native_collection
        self._algorithms = [Algorithm.create_from_native_handle(algo) for algo in self._native_collection.to_list()]

    def __iter__(self):
        """Iterate over all algorithms in the collection."""
        return iter(self._algorithms)

    def __len__(self):
        """Return the number of algorithms in the collection."""
        return len(self._algorithms)

    def __getitem__(self, index: int) -> Algorithm:
        """Get an algorithm by index."""
        return self._algorithms[index]

    def get_by_collective(self, collective: str):
        """Get all algorithms for a specific collective operation."""
        return [algo for algo in self._algorithms if algo.collective == collective]

    def register_algorithm(self, collective: str, algo_name: str, algorithm: Algorithm):
        """Register an algorithm for a collective operation."""
        self._native_collection.register_algorithm(collective, algo_name, algorithm._algorithm)
        self._algorithms.append(algorithm)